klippa/src/cmap.rs

```diff
@@ -0,0 +1,1105 @@
+            }
+        } else if format == 12 {
+            let Some(unicodes_set) =
+                unicodes_cache.set_for(*rec_idx, &subtable, plan.font_num_glyphs)
+            else {
+                continue;
+            };
+            let cmap12_subset_unicodes =
+                IntSet::from_iter(plan.unicodes.iter().filter(|v| unicodes_set.contains(*v)));
+
+            if can_drop_format12(
+                record,
+                &cmap12_subset_unicodes,
+                cmap,
+                retained_encoding_records,
+                &mut unicodes_cache,
+                &plan.unicodes,
+                plan.font_num_glyphs,
+            ) {
+                continue;
+            }
+
+            let cp_to_new_gid_list: Vec<(u32, GlyphId)> = plan
+                .unicode_to_new_gid_list
+                .iter()
+                .filter(|&(cp, _gid)| cmap12_subset_unicodes.contains(*cp))
+                .map(|(cp, gid)| (*cp, *gid))
+                .collect();
+            format12_objidx =
+                serialize_encoding_record(record, &subtable, s, &cp_to_new_gid_list, plan)?;
+        } else if format == 14 {
+            serialize_encoding_record(record, &subtable, s, &plan.unicode_to_new_gid_list, plan)?;
+        }
+    }
+
+    let num_retained_records = (s.length() - 4) / EncodingRecord::RAW_BYTE_LEN;
+    s.check_assign::<u16>(
+        2,
+        num_retained_records,
+        SerializeErrorFlags::SERIALIZE_ERROR_INT_OVERFLOW,
+    );
+
+    // Fail if format 4 was dropped and there is no cmap12.
+    if drop_format_4 && format12_objidx.is_none() {
+        return Err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+    }
+
+    Ok(())
+}
+
+fn serialize_encoding_record(
+    record: &EncodingRecord,
+    cmap_subtable: &CmapSubtable,
+    s: &mut Serializer,
+    cp_to_new_gid_list: &[(u32, GlyphId)],
+    plan: &Plan,
+) -> Result<Option<ObjIdx>, SerializeErrorFlags> {
+    let snap = s.snapshot();
+    s.embed(record.platform_id())?;
+    s.embed(record.encoding_id())?;
+    let offset_pos = s.embed(Offset32::new(0))?;
+
+    s.push()?;
+    let init_len = s.length();
+    cmap_subtable.serialize(s, plan, cp_to_new_gid_list)?;
+    let mut obj_idx = None;
+    if s.length() > init_len {
+        obj_idx = s.pop_pack(true);
+        if obj_idx.is_none() {
+            s.revert_snapshot(snap);
+            return Err(s.error());
+        }
+    } else {
+        s.pop_discard();
+        s.revert_snapshot(snap);
+        return Ok(obj_idx);
+    }
+
+    s.add_link(
+        offset_pos..offset_pos + 4,
+        obj_idx.unwrap(),
+        OffsetWhence::Head,
+        0,
+        false,
+    )?;
+    Ok(obj_idx)
+}
+
+fn is_gid_consecutive(
+    end_char_code: u32,
+    start_char_code: u32,
+    gid: GlyphId,
+    cp: u32,
+    new_gid: GlyphId,
+) -> bool {
+    cp - 1 == end_char_code && new_gid.to_u32() == gid.to_u32() + (cp - start_char_code)
```

Comment:
do we know this will never be called with `cp == 0`?

Reply:
yes I think so